Compile time BuildInfo class generator

MarcinOrlowski opened the pull request:
This is my take on simplifying project information like version etc that are currently hardcoded and spread among multiple files (i.e. version is in `Main.java` and in `gradle.properties`). This PR introduces `genBuildInfo` task that generates `BuildInfo.java` file with information read from `gradle.properties` (like version) and also compile time details like current stamp, branch name or last commit hash (this info is also added to About scroller in this PR), so at the end of the day the only source of project information to maintain would be `gradle.properties`. I also did minor gradle cleanup.

@R3dst0ne @mbaillif @davidhutchens @maehne 

BLOCKED_BY #888

MarcinOrlowski commented:
BTW: do we need `jpackage` task?

BFH-ktt1 commented:
> BTW: do we need `jpackage` task?

I would like to keep it. Maybe in the near future we can supres it, but for the moment I use it regularly.

MarcinOrlowski commented:
May I rename it to `createAll` as this is basically what it does?

R3dst0ne commented on build.gradle.kts:
```diff
@@ -45,31 +72,40 @@ java {
 task<Jar>("sourcesJar") {
   group = "build"
-  description = "Creates a source jar archive."
+  description = "Creates a source JAR archive file."
```
`source JAR archive` is a kind of file, so the "file" at the end of the line is redundant.

R3dst0ne commented on build.gradle.kts:
```diff
@@ -78,40 +114,42 @@ extra.apply {
 tasks.register("createDeb") {
   group = "build"
-  description = "Makes the Linux platform specific packages"
+  description = "Makes DEB Linux package file."
```
`DEB package` is a kind of file, so the "file" at the end of the line is redundant.

R3dst0ne commented on build.gradle.kts:
```diff
@@ -126,18 +164,23 @@ tasks.register("createDeb") {
 tasks.register("createRpm") {
   group = "build"
-  description = "Makes the Linux platform specific packages"
+  description = "Makes RPM Linux package file."
```
`RPM package` is a kind of file, so the "file" at the end of the line is redundant.

R3dst0ne commented on build.gradle.kts:
```diff
@@ -150,17 +193,22 @@ tasks.register("createRpm") {
 tasks.register("createMsi") {
   group = "build"
-  description = "Makes the Windows platform specific package"
+  description = "Makes the Windows platform specific package."
```
Here, you didn't add "file" to the end. Consider making it consistent.
(The best option is to remove the redundant "file"s.)

R3dst0ne commented on build.gradle.kts:
```diff
@@ -174,25 +222,30 @@ tasks.register("createMsi") {
       val process1 = processBuilder1.start()
       if (process1.waitFor() != 0) {
-        throw GradleException("Error while creating the .msi package")
+        throw GradleException("Error while creating the MSI package file.")
```
`MSI package` is a kind of file, so the "file" at the end of the line is redundant.

R3dst0ne commented on build.gradle.kts:
```diff
@@ -243,15 +296,20 @@ tasks.register("createApp") {
 tasks.register("createDmg") {
   group = "build"
-  description = "Makes the Mac dmg package"
+  description = "Makes the Mac DMG package file."
```
`DMG package` is a kind of file, so the "file" at the end of the line is redundant.

davidhutchens commented:
Removing the jpackage task would require adding descriptions for all the individual tasks for each platform to the docs/developer.md. Having a single "build the thing for my system" task allows us to keep the docs/developers.md simple.

Renaming it to createAll, createPackages, or some such, is an improvement. It is usually better to name a task for what it does for the user rather than for how it is implemented. If you rename it, also change docs/developer.md to match.

davidhutchens commented:
While you are working on the build file, could you also change the "import from java.util.*" to just those parts we actually use? I think it is just "java.util.Date" with your changes to listOf.

MarcinOrlowski commented:
I'll rename that task in separate PR then.

davidhutchens commented:
The generateBuildInfoClassFile always runs. This forces a recompile each time any other task is run. To observe, try "./gradlew compileJava" and then repeat the command. It should report no new tasks have executed since all should be up-to-date. It reports 2 executed.

I suspect you need to specify the input and output files for the task.

MarcinOrlowski commented:
Yup, task is dumb ATM. I am thinking about turning it info a regular function and then make `build` task call when it runs, instead of having a dependency. Alternatively `generateBuildInfoClassFile` should depend on the same `sourceSet` the `build` task is using. Unless I miss something obvious it could be sufficient (the task can stay a standalone if needed).

MarcinOrlowski commented:
Incremental builds should now work as expected.

davidhutchens commented on build.gradle.kts:
```diff
@@ -355,6 +355,9 @@ tasks.register("generateBuildInfoClassFile") {
   val projectInfoDir = "${buildDir}/generated/logisim/java/com/cburch/logisim/generated"
   val projectInfoFile = "${projectInfoDir}/BuildInfo.java"
 
+  inputs.files(fileTree("src/main")).withPropertyName("sourceFiles")
```
This doesn't seem to force a rebuild if you change the version in gradle.properties. I suggest something like: 

  inputs.dir("$projectDir/src")
  inputs.dir("$projectDir/support")
  inputs.dir("$projectDir/artwork")
  inputs.files("$projectDir/gradle.properties", "$projectDir/README.md", "$projectDir/LICENSE.md")

That seems to force a new timestamp when something going into the distribution packages changes. Of course, I still might have omitted something.

I note that this also hard-codes things. I don't currently have a solution that doesn't.

MarcinOrlowski replied:
BTW: I wonder if we should not fix the dependency for `compileJava` to contain:
```
inputs.files("$projectDir/gradle.properties", "$projectDir/README.md", "$projectDir/LICENSE.md")
```
because in fact it means we need to rebuild the app, not to generate a class (which is a consequence of building, not a reason for it).

davidhutchens replied:
We would still need the dependsOn("genBuildInfo") since it must run before compileJava. I tried removing that and genBuildInfo doesn't even run when if I change gradle.properties and invoke a higher task such as createApp. Adding the inputs.files you suggest doesn't change that. Apparently, if genBuildInfo does run, and compileJava is requested, it actually recompiles because a source (the buildinfo) has changed. So I don't think it would hurt to add the line you suggest (as long as we leave the dependsOn), but I don't think it is technically necessary.

MarcinOrlowski commented:
Making draft as I want #888 to land first.

MarcinOrlowski commented:
Right, touching `gradle.properties` must trigger the build.  Good spot. Fixed.

davidhutchens commented:
Using a version with a "-alpha1" added on the Mac causes jpackage to fail with the message:

Bundler Mac Application Image skipped because of a configuration problem: "Version [3.5.0-alpha1] contains invalid component [0-alpha1]" 
Advice to fix: Set a compatible 'app-version' value. Valid versions are one to three integers separated by dots.

davidhutchens commented:
Changing the code in the create* tasks so they fail with an exception when the platform can't do the task is a good idea. I initially tried to do that, but it caused the jpackage task to always fail as it depends on each of them. That is why each of them was simply a no-operation when the platform couldn't build the target.

However, your attempt caused me to have another look. It turns out we can change the jpackage task to replace the dependsOn line with:

  if (OperatingSystem.current().isLinux) {
    dependsOn("createDeb", "createRpm")
  }
  if (OperatingSystem.current().isWindows) {
    dependsOn("createMsi")
  }
  if (OperatingSystem.current().isMacOsX) {
    dependsOn("createDmg")
  }

That makes the dependency be platform specific! So you need to make that change to keep the jpackage task working.

MarcinOrlowski commented:
I fixed `jbuilder` as you suggested as this is IMHO proper way -> the task should fail when invoked on unsupported target platform on that host and not be invoked by "platform agnostic" tasks.

MarcinOrlowski commented:
Can you please try version without a dash, i.e. `3.6.0beta1`?

davidhutchens commented:
Running just the jdk jpackage command (not using our build system at all) with all our normal options and `--version 3.5.0beta1` gives:

Bundler Mac Application Image skipped because of a configuration problem: "Version [3.5.0beta1] contains invalid component [0beta1]" 
Advice to fix: Set a compatible 'app-version' value. Valid versions are one to three integers separated by dots.

MarcinOrlowski commented:
OK, thanks for the feedback. I will take another round on versioning then.

davidhutchens commented on build.gradle.kts:
```diff
@@ -348,7 +352,7 @@ tasks.register("createDmg") {
           "--app-image", ext.get("appDirName") as String,
           "--name", project.name,
-          "--app-version", ext.get("appVersion") as String,
+          "--app-version", ext.get("appVersionShort") as String,
```
It turns out that you did not need to change this line. But you do need to add this appVersionShort line to the createApp task at line 289. The failure occurs in createApp which is still getting its value from the sharedParameters. Once createApp succeeds, you can add the suffix to the .dmg name when running createDmg. Apparently jpackage only checks the version format when building the app (which needs to match Apple's version formatting). That way the app's Info.plist has the "real" version in a.b.c form and the Logisim-evolution-3.5.0-beta1.dmg file contains the suffix after the version.

I also realized in testing this, that jpackage will use the last setting of the parameter. So if it is set in the sharedParameters, it can be overridden by also including it in the list that is added in createApp. I'm not sure if that is a good idea. It simplifies our code; but at some time in the future, jpackage may decide to detect it has been set twice and complain. What do you think?

MarcinOrlowski replied:
I updated the macOS related tasks according to your notes. Please let me know if all is fine.

> jpackage may decide to detect it has been set twice and complain. What do you think?

I would not worry about that. Such detection would be of no sense really. In fact many CLI tools allow setting the same flag multiple times and the last one wins because that also simplifies constructing command line invocation strings. Anyway, we can worry once that become a real problem.

davidhutchens commented:
This now works on macOS.

MarcinOrlowski commented:
I did more changes the the building script. First, I refactored use of `ProcessBuilder()` to reduce LoC, so there's now a dedicated helper function for it. Also I reworked `createDistDir` task to no longer fail if if there's more than one JAR in `${buildDir}/libs` but remove all files but the one expected (it will post `warn`inings when this happen).

davidhutchens commented on build.gradle.kts:
```diff
@@ -104,54 +104,130 @@ extra.apply {
+  set("sharedParameters", params)
+
+  // Linux (DEB/RPM) specific settings for jpackage.
+  params.addAll(arrayOf(
```
You set the sharedParameters to params which is good. Then you changed params by adding the Linux specific things. This changes the object that sharedParameters uses. The result is that sharedParameters now has Linux specific options. This causes createApp (and I assume createMsi) to fail. The original code created a new ArrayList object for the Linux things rather than adding to the old.

MarcinOrlowski replied:
I already fixed that locally. This approach was fine as we were constructing `ArrayList` to wrap params. Once I dropped that it make no sense. Just add `.clone()` to line 128:

    set("sharedParameters", params.clone())

davidhutchens replied:
Yes, clone() works.

davidhutchens commented:
I like the change to remove any old files from build/libs and continue with the build. I was thinking about doing that.

davidhutchens reviewed (commented):
Has any of this been tested on Windows?

davidhutchens commented on build.gradle.kts:
```diff
@@ -104,54 +104,130 @@ extra.apply {
+      "--main-jar", shadowJarFilename,
       "--app-version", version,
-      "--copyright", "\"Copyright ©2001–${SimpleDateFormat("yyyy").format(Date())} ${project.name} developers\"",
+      "--copyright", "\"${copyrights}\"", // Must be quoted for shell happines.
```
When you added the quotes to the copyright string, were you addressing a specific failure? Because the result is that the quotes are passed right through to the Info.plist file in the Mac app. There is nothing stripping them out. They are not needed in the Info.plist file since it uses xml style tags: <string>stuff</string>

I also notice that you named the function runShellCommand. However, there is no shell in the process. processBuilder passes the array of parameters directly to a Unix system call (one of the exec calls) without using a shell for anything. I'm not positive what happens on Windows, but I suspect there is no shell equivalent there either.

That is why putting quote marks around the jpackage command causes failure on my Mac (and I strongly suspect, on all Unix systems including Linux).

MarcinOrlowski replied:
> were you addressing a specific failure?

Yes. 

> They are not needed in the Info.plist file since it uses xml style tags

But quotes are string wrappers only, not a part of it. Should not pass IIRC. I can double check that on macOS/Windows tomorrow.

Windows weren't tested yet.

davidhutchens commented:
createAll fails because dependsOn cannot be called after a task starts. The ifs cannot be inside doFirst.

MarcinOrlowski commented:
Indeed. I moved it out of configuration phase, which made no sense. Fixed. Thanks.

MarcinOrlowski commented:
Windows [builder](https://github.com/MarcinOrlowski/logisim-evolution/runs/3381223818?check_suite_focus=true) seems working too, so  ready to review.

davidhutchens commented:
I have two minor issues.

The quote marks are still inserted into the Info.plist on the Mac. I still do not understand why they are necessary (and no other application on my system puts quotes there). There is no shell in the build process. They are not necessary for the code to work, as far as I can tell. The comment on the runShellCommand:
 * Helper method that concatenates params argument list into valid
 * shell command and executes it using ProcessBuilder().

is incorrect. It does not concatenate params, nor does it use a shell. ProcessBuilder does not internally concatenate the parameters or use a shell either. The code is correct as written. The comment is wrong. (Well, technically the code concatenates the params for the logger; but that doesn't change what it is actually doing).

The second issue is that the outputs.file for createDeb is incorrect. The format for the file has an underscore between the project.name and the version. targetFilePathBase has a dash. To see the minor issue, do createDeb, remove the dist/*.deb file, and do createDeb. It should recreate it; but since the outputs.file is incorrect, it will not.

Neither of these is a critical problem.

BFH-ktt1 commented:
@MarcinOrlowski : This part is out of my competences, hence it does not make much sense for me to review.

MarcinOrlowski commented:
I corrected the docs and renamed method to just `runCommand()`.

MarcinOrlowski commented:
Fixed `createDeb` task.

MarcinOrlowski commented:
I removed the quotation of copyright string and Deb and Rpm tasks seem work correctly. I still wonder what's the difference with `jPackageCmd` that it needs conditional quoting. Are you sure that a) you have to quote even if there's space in the path? b) can't quote if there's no space (that'd be odd)?

davidhutchens commented:
Thank you.

If you look at my last comments in #829, you will see that I said that putting quotes around the command looked wrong, and in fact fails on my Mac, and I figured that it was always wrong. Nobody ever tested that on Linux and Windows, but I'm more convinced of that now. So, no, I am not sure those quotes should be there. Quite the opposite.

What I actually said is that not quoting always works, even with spaces in the command, specifically: "It does work with a space in the path if I do not add quotes."

I suggest we simply remove that test and never put quotes around the command. It would be nice to test that on Linux and Windows, but I am now very sure quotes are always wrong.

davidhutchens reviewed (approved):
I approve this PR. Thanks, @MarcinOrlowski.

MarcinOrlowski commented:
I also checked on Linux. I copied JDK to folder with spaces and then copied sources to different folder with spaces and then succesfuly built DEB and RPM w/o any special quotes.
